modyn/supervisor/internal/triggers/cost/costtrigger.py

```diff
@@ -0,0 +1,148 @@
+from __future__ import annotations
+
+import logging
+from abc import abstractmethod
+from collections.abc import Generator
+
+from typing_extensions import override
+
+from modyn.config.schema.pipeline.trigger.cost.cost import CostTriggerConfig
+from modyn.supervisor.internal.triggers.cost.cost_tracker import CostTracker
+from modyn.supervisor.internal.triggers.cost.incorporation_latency_tracker import (
+    IncorporationLatencyTracker,
+)
+from modyn.supervisor.internal.triggers.models import (
+    CostAwareTriggerEvalLog,
+    TriggerPolicyEvaluationLog,
+)
+from modyn.supervisor.internal.triggers.trigger import Trigger, TriggerContext
+
+logger = logging.getLogger(__name__)
+
+
+class CostTrigger(Trigger):
+    """Triggers when a certain number of data points have been used."""
+
+    def __init__(self, config: CostTriggerConfig):
+        self.config = config
+        self.context: TriggerContext | None = None
+
+        self._sample_left_until_detection = config.detection_interval_data_points
+        self._triggered_once = False
+        self._previous_batch_end_time: int | None = None
+        self._leftover_data: list[tuple[int, int]] = []
+        """Stores data that was not processed in the last inform call because
+        the detection interval was not filled."""
+
+        self._unincorporated_samples = 0
+        self._cost_tracker = CostTracker(config.cost_tracking_window_size)
+        self._incorporation_latency_tracker = IncorporationLatencyTracker()
+        """Maintains the regret metric and the cumulative regret latency,
+        semantics are defined by the subclass."""
+
+    @override
+    def init_trigger(self, context: TriggerContext) -> None:
+        self.context = context
+
+    def inform(
+        self,
+        new_data: list[tuple[int, int, int]],
+        log: TriggerPolicyEvaluationLog | None = None,
+    ) -> Generator[int, None, None]:
+        new_key_ts = self._leftover_data + [(key, timestamp) for key, timestamp, _ in new_data]
+        # reappending the leftover data to the new data requires incrementing the sample left until detection
```

Comment:
These 20 lines are duplicated across all policies that use discretized time units for detection. I think this is okay. An abstraction for this with polymorphic delegation functions would make the code quite uncomprehensive, as every trigger has some special data structures that need to be updated within this section.

Reply:
** Resolved via BatchedTrigger